cruise-control/src/main/scala/com/linkedin/kafka/cruisecontrol/executor/ExecutorUtils.scala

```diff
@@ -85,38 +87,87 @@ object ExecutorUtils {
                 addTask = false
                 Seq.empty
               } else {
-                if (!currentReplicaAssignment.contains(destinationBroker)) {
-                  if (currentReplicaAssignment.contains(sourceBroker)) {
-                    // this is a normal movement.
-                    if (destinationBroker != null)
-                      (currentReplicaAssignment :+ destinationBroker.toInt).filter(_ != sourceBroker) 
-                    else
-                      currentReplicaAssignment.filter(_ != sourceBroker)
-                  } else {
-                    // The replica list should have at least one of the source broker or destination broker.
-                    throw new RuntimeException(s"Broker $sourceBroker is not a replica of [$topic, $partition].")
+                if (task.proposal.balancingAction() == BalancingAction.REPLICA_MOVEMENT) {
+                  if (currentReplicaAssignment.contains(destinationBroker) && !currentReplicaAssignment.contains(sourceBroker)) {
+                    // Reassignment is done already.
+                    addTask = false
                   }
-                } else {
-                  if (currentReplicaAssignment.contains(sourceBroker)) {
-                    // The destination broker is already in the list, we just need to filter out the source broker if
-                    // it exists.
-                    currentReplicaAssignment.filter(_ != sourceBroker)
-                  } else {
-                    // If the source broker is no longer in the list, just do not add the task to the reassignment.
+                  // Get new replicas for replica movement.
```

Comment:
Should this be in the else branch? After set addTask=false, we can just return Seq.empty. ditto below.